src/UI/Implementation/Component/Table/Data.php

```diff
@@ -0,0 +1,304 @@
+use ILIAS\UI\Implementation\Component\SignalGeneratorInterface;
+use ILIAS\UI\Component\Signal;
+use ILIAS\UI\Implementation\Component\JavaScriptBindable;
+use ILIAS\UI\Component\JavaScriptBindable as JSBindable;
+use ILIAS\Data\Factory as DataFactory;
+use ILIAS\Data\Order;
+use ILIAS\Data\Range;
+
+class Data extends Table implements T\Data, JSBindable
+{
+    use JavaScriptBindable;
+
+    protected string $title;
+    protected DataRetrieval $data_retrieval;
+
+    /**
+     * @var array <string, Column>
+     */
+    protected $columns = [];
+
+    /**
+     * @var array <string, Action>
+     */
+    protected $actions = [];
+
+    protected Signal $multi_action_signal;
+    protected Signal $selection_signal;
+    protected ?ServerRequestInterface $request = null;
+    protected DataFactory $data_factory;
+
+    protected int $number_of_rows = 800;
+    protected array $selected_optional_column_ids = [];
+    protected Range $range;
+    protected Order $order;
+    protected ?array $filter = null;
+    protected ?array $additional_parameters = null;
+
+    public function __construct(
+        SignalGeneratorInterface $signal_generator,
+        DataFactory $data_factory,
+        string $title,
+        array $columns,
+        T\DataRetrieval $data_retrieval
+    ) {
+        parent::__construct($title);
+        $this->multi_action_signal = $signal_generator->create();
+        $this->selection_signal = $signal_generator->create();
+        $this->data_factory = $data_factory;
+        $this->data_retrieval = $data_retrieval;
+        $this->setEnumeratedColumns($columns);
+        $this->initializeVisibleColumns();
+        $this->initializeViewDefaults();
+    }
+
+    protected function setEnumeratedColumns(array $columns): void
+    {
+        if (count($columns) === 0) {
+            throw new \InvalidArgumentException('cannot construct a table without columns.');
+        }
+        $counter = 0;
+        foreach ($columns as $id => $column) {
+            if (!is_a($column, T\Column\Column::class)) {
+                throw new \InvalidArgumentException($id . ' is not a column.');
+            }
+            $this->columns[$id] = $column->withIndex($counter);
+            $counter++;
+        }
+    }
+    protected function initializeVisibleColumns(): void
+    {
+        $this->selected_optional_column_ids =  array_keys(
+            array_filter(
+                $this->getColumns(),
+                fn ($c) => $c->isInitiallyVisible()
+            )
+        );
+    }
+
+    protected function initializeViewDefaults(): void
+    {
+        $this->range = $this->data_factory->range(0, $this->number_of_rows);
+
+        $sortable_visible_cols = array_filter(
+            $this->getVisibleColumns(),
+            fn ($c) => $c->isSortable()
+        );
+
+        $order_by = current(array_keys($sortable_visible_cols));
+        $this->order = $this->data_factory->order($order_by, \ILIAS\Data\Order::ASC);
```

Comment:
This currently fails in when
a) you build a table without any sortable column (`Argument #1 ($subject) must be of type string, bool given`)
b) your array of columns doesn't have strings as array indices (`Argument #1 ($subject) must be of type string, int given`)
I think both should work :-)